### ISSUE TITLE
Improper nonce handling in Go

### ISSUE THREAD

**Issue by dnkolegov** [symbolicsoft/noiseexplorer#1]
There are two issues with nonce handling in Go:
1. Nonces are 32 bits long. The Noise [spec](https://noiseprotocol.org/noise.html#the-cipherstate-object) requires 64-bit nonce.

```
type cipherstate struct {
	k [32]byte
	n uint32
}
```

2. Nonce overflow.
The [spec](https://noiseprotocol.org/noise.html#the-cipherstate-object) requires returning an error if the maximum number is reached. This is the same [issue](https://github.com/flynn/noise/security/advisories/GHSA-g9mp-8g3h-3c5c) as in the flynn package but more realistic due to 32-bit length.

```
func incrementNonce(n uint32) uint32 {
	return n + 1
}

func encryptWithAd(cs *cipherstate, ad []byte, plaintext []byte) (*cipherstate, []byte) {
	e := encrypt(cs.k, cs.n, ad, plaintext)
	cs = setNonce(cs, incrementNonce(cs.n))
	return cs, e
}
```

**Comment by nadimkobeissi** [comment]
@dnkolegov Thanks for reporting this. I'll get it fixed soon.

**Comment by nadimkobeissi** [comment]
Fixed in [Noise Explorer 1.0.3](https://source.symbolic.software/noiseexplorer/noiseexplorer/-/releases/v1.0.3). Thanks again.
